### PR TITLE
VFB-170: Add profiles table to store user information

### PR DESCRIPTION
**Pull request by minalui**
## What's changed
- Add profiles table in public database to store profile information of users
- Change user edge functions into database functions
- Use profile table to retrieve user role instead of app_metadata
- Create new card for create user form for first, last name and telephone number

## Screenshots / Videos
| Before     | After      |
|------------|------------|
| image | image |

## Checklist
- [x] The ticket is up-to-date - Please document any deviations from the original approach if there is any.
- [x] I have documented the testing steps for QA
- [x] I have self-reviewed this PR
- [x] Make sure you've verified it works via `npm run dev`
- [x] Make sure you've verified it works via `npm run build` and `npm run start`
- [x] Make sure you've fixed all linting problems with `npm run lint_fix`
- [x] Make sure you've tested via `npm run test:e2e`

If you have made any changes to the database...
  - [x] The migration files are up-to-date with my final set up (`npx supabase db diff -f <name_of_migration>` should create nothing at this point)
  - [x] I have updated the typescript definitions for the database with `db:local:generate_types`
  - [x] I have modified the seed in `supabase/seed/seed.mts` if appropriate
  - [x] If I have modified the seed, I have also generated the seed with `npm run db:generate_seed` 
  - [x] With my final set up, I can run `npm run dev:reset_supabase` without any errors.
  - Does this require resetting the deployed database? - YES (remove as appropriate)

### CODE REVIEW

**Review by minalui — COMMENTED** [review]
Self-review

File: src/common/fetchUserRole.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+): Promise<Database["public"]["Enums"]["role"]> => {
+    const { data, error } = await supabase
+        .from("profiles")
+        .select("role")
+        .eq("primary_key", userId)
+        .single();
+
+    if (error) {
+        if (logError) {
+            const logId = logError(error);
```
Comment (minalui) [inline src/common/fetchUserRole.ts:18]: There should be a error message submitted with logError?

File: src/server/adminCreateUser.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    if (data) {
+        const { error: createRoleError } = await supabase.from("profiles").insert({
+            primary_key: data.user?.id,
+            role: userDetails.role,
+            first_name: userDetails.firstName,
+            last_name: userDetails.lastName,
+            telephone_number: userDetails.telephoneNumber,
+        });
+        if (createRoleError) {
+            console.log(createRoleError);
```
Comment (minalui) [inline src/server/adminCreateUser.ts:52]: This should log an error on CloudWatch

File: src/server/adminCreateUser.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        }
+        console.log(`Created a profile for ${userDetails.role} user: ${userDetails.email}`);
```
Comment (minalui) [inline src/server/adminCreateUser.ts:54]: This should logInfo

File: src/server/adminDeleteUser.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    const adminAuthClient = getSupabaseAdminAuthClient();
+    const { error } = await adminAuthClient.deleteUser(userId);
+
+    if (error) {
+        return {
+            error: { AuthError: `error deleting user: ${userId}` },
+        };
+    }
+
```
Comment (minalui) [inline src/server/adminDeleteUser.ts:27]: Add a logInfo?

File: src/app/admin/page.tsx [hunk 5]
```diff
@@ -21,25 +23,27 @@ export interface UserRow {
...
+        const userRole = await fetchUserRole(user.id, (error) => {
+            return logErrorReturnLogId("Failed to fetch user role", { error });
+        });
 
-    return users.map((user: User) => {
         return {
             id: user.id,
             email: user.email ?? "-",
-            userRole: user.app_metadata.role ?? "-",
+            userRole: userRole ?? "-",
```
Comment (rickaw) [inline src/app/admin/page.tsx:41]:
```suggestion
            userRole: userRole,
```

File: src/app/errorClasses.ts [hunk 6]
```diff
@@ -1,21 +1,25 @@
 export class DatabaseError extends Error {
-    constructor(errorCategory: DatabaseOperation, faultyArea = "", logId: string) {
+    constructor(errorCategory: DatabaseOperation, faultyArea = "", logId?: string) {
         super(
             `An error occurred completing the ${errorCategory} operation` +
                 (faultyArea !== "" ? ` (${faultyArea})` : "") +
                 ". Please try again." +
-                `Log ID: ${logId}`
+                logId
+                ? `Log ID: ${logId}`
```
Comment (rickaw) [inline src/app/errorClasses.ts:8]: (same in the other places)

```suggestion
                ? ` Log ID: ${logId}`
```

File: src/app/parcels/ParcelsPage.tsx [hunk 7]
```diff
@@ -434,31 +434,40 @@ const ParcelsPage: React.FC<{}> = () => {
...
-        if (parcelId) {
-            setSelectedParcelId(parcelId);
-            setModalIsOpen(true);
-        }
-    }, [parcelId]);
-
-    useEffect(() => {
-        const fetchAndSetClientIdForSelectedParcel = async (): Promise<void> => {
+    const fetchAndSetClientIdForSelectedParcel = async (): Promise<void> => {
+        if (!parcelId) {
```
Comment (rickaw) [inline src/app/parcels/ParcelsPage.tsx:438]: why do we check this? (the logic around here will change from Michael's bug fix soon though https://github.com/NorwoodAndBrixtonFoodbank/nbf-website/pull/226)

File: src/app/parcels/ParcelsPage.tsx [hunk 8]
```diff
@@ -434,31 +434,40 @@ const ParcelsPage: React.FC<{}> = () => {
...
+            setSelectedParcelId(parcelId);
+            setModalIsOpen(true);
+        }
+    }, [parcelId]);
+
+    useEffect(() => {
+        if (parcelId) {
+            void fetchAndSetClientIdForSelectedParcel();
+        }
         setClientIdForSelectedParcel(null);
```
Comment (rickaw) [inline src/app/parcels/ParcelsPage.tsx:470]: We should probably set client id to null first right?
```suggestion
        setClientIdForSelectedParcel(null);
        if (parcelId) {
            void fetchAndSetClientIdForSelectedParcel();
        }
```

**Review by rickaw — CHANGES_REQUESTED** [review]

**Comment by rickaw** [comment]
Love how the existing edge functions are written really neatly with server actions!

File: src/server/adminCreateUser.ts [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+    const adminAuthClient = getSupabaseAdminAuthClient();
+    const { data, error } = await adminAuthClient.createUser({
+        email: userDetails.email,
+        password: userDetails.password,
+        email_confirm: true,
+    });
+
+    if (error) {
+        return {
```
[the remark this answers is not in the document]
Comment (minalui) [reply src/server/adminCreateUser.ts:40]: The errors are currently returned to the component that called the form and logged there. I have added some logs for updated password and update profile.

File: src/app/parcels/ParcelsPage.tsx (hunk 7 above)
Comment (minalui) [reply src/app/parcels/ParcelsPage.tsx:438]: I think it kept trying to run this function when no parcels are being selected and error messages kept coming up

**Review by rickaw — CHANGES_REQUESTED** [review]

File: src/supabaseAdminAuthClient.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    if (!supabaseKey) {
+        throw new Error("Supabase key is not set");
+    }
+
+    const supabase = createClient(supabaseUrl, supabaseKey, {
+        auth: {
+            autoRefreshToken: false,
+            persistSession: false,
+        },
+    });
```
[the remark this answers is not in the document]
Comment (minalui) [reply src/supabaseAdminAuthClient.ts:21]: Methods under the supabase.auth.admin namespace requires a service_role key

File: src/app/admin/createUser/CreateUserForm.tsx [hunk 11]
```diff
@@ -63,14 +73,17 @@ const CreateUserForm: React.FC<{}> = () => {
...
 
         if (error) {
             setSubmitError(Errors.external);
             setSubmitDisabled(false);
             setCreatedUser(null);
 
-            const logId = await logErrorReturnLogId("Error with insert: User", error);
+            const logId = await logErrorReturnLogId(
+                `Error with insert: User ${fields.firstName} ${fields.lastName}`,
+                error
```
Comment (rickaw) [inline src/app/admin/createUser/CreateUserForm.tsx:85]:
```suggestion
                { error }
```

File: src/app/admin/deleteUser/DeleteUserDialog.tsx [hunk 12]
```diff
@@ -50,7 +54,10 @@ const DeleteUserDialog: React.FC<Props> = (props) => {
                 success: false,
                 message: <>Delete User Operation Failed</>,
             });
-            const logId = await logErrorReturnLogId("Error with delete: User", error);
+            const logId = await logErrorReturnLogId(
+                `Error with delete: User ${props.userToDelete.email}`,
+                error
```
Comment (rickaw) [inline src/app/admin/deleteUser/DeleteUserDialog.tsx:59]:
```suggestion
                { error }
```

File: src/app/admin/manageUser/UpdateUserProfile.ts [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+        .from("profiles")
+        .update({
+            role: userDetails.role,
+        })
+        .eq("primary_key", userDetails.userId)
+        .select()
+        .single();
+
+    if (error) {
+        console.log(error);
```
Comment (rickaw) [inline src/app/admin/manageUser/UpdateUserProfile.ts:26]: should we log properly here?

File: src/app/roles.tsx [hunk 14]
```diff
@@ -1,43 +1,44 @@
...
 interface Props {
     children: React.ReactNode;
 }
 
-type Roles = DatabaseEnums["role"] | "";
-
-type RolesInterface = {
-    [key in Roles]: string[];
-};
+export type Roles = DatabaseRoles;
```
Comment (rickaw) [inline src/app/roles.tsx:10]: If we're just redeining `DabaseRoles`, can we just use `DatabaseRoles` and not `Roles` at all?

File: src/databaseUtils.ts [hunk 15]
```diff
@@ -18,6 +18,7 @@ export type UpdateSchema = {
 };
 
 export type DatabaseEnums = Database["public"]["Enums"];
+export type DatabaseRoles = DatabaseEnums["role"];
```
Comment (rickaw) [inline src/databaseUtils.ts:21]:
```suggestion
export type UserRole = DatabaseEnums["role"];
```

**Review by rickaw — APPROVED** [review]
Some small comments but feel free to merge after they're addressed!

File: src/app/admin/manageUser/UpdateUserProfile.ts (hunk 13 above)
Comment (minalui) [reply src/app/admin/manageUser/UpdateUserProfile.ts:26]: Logged in editUserForm, I've just removed the console log now

File: src/app/roles.tsx (hunk 14 above)
Comment (minalui) [reply src/app/roles.tsx:10]: Yes

**Pull request merged**